src/OrbitGl/TimerTrack.cpp

```diff
@@ -110,151 +251,34 @@ void TimerTrack::UpdatePrimitives(uint64_t min_tick, uint64_t max_tick,
-          }
-        }
-
-        uint64_t function_id = timer_info.function_id();
-
-        bool is_selected = &text_box == selected_textbox;
-        bool is_highlighted = !is_selected &&
-                              function_id != orbit_grpc_protos::kInvalidFunctionId &&
-                              function_id == highlighted_function_id;
-
-        static const Color kHighlightColor(100, 181, 246, 255);
-        Color color = is_highlighted ? kHighlightColor : GetTimerColor(timer_info, is_selected);
-
-        bool is_visible_width = elapsed_us * inv_time_window * canvas->GetWidth() > 1;
-
-        if (is_visible_width) {
-          WorldXInfo world_x_info_left_overlap =
-              ToWorldX(start_us, start_or_prev_end_us, inv_time_window, world_start_x, world_width);
-
-          WorldXInfo world_x_info_right_overlap =
-              ToWorldX(end_or_next_start_us, end_us, inv_time_window, world_start_x, world_width);
-
-          Vec3 top_left(world_x_info_left_overlap.world_x_start, world_timer_y + box_height, z);
-          Vec3 bottom_left(
-              world_x_info_left_overlap.world_x_start + world_x_info_left_overlap.world_x_width,
-              world_timer_y, z);
-          Vec3 top_right(world_x_info_right_overlap.world_x_start, world_timer_y + box_height, z);
-          Vec3 bottom_right(
-              world_x_info_right_overlap.world_x_start + world_x_info_right_overlap.world_x_width,
-              world_timer_y, z);
-          batcher->AddShadedTrapezium(
-              top_left, bottom_left, bottom_right, top_right, color,
-              std::make_unique<PickingUserData>(
-                  &text_box, [&](PickingId id) { return this->GetBoxTooltip(id); }));
-
-        } else {
-          auto user_data = std::make_unique<PickingUserData>(
-              &text_box, [&](PickingId id) { return this->GetBoxTooltip(id); });
-
-          WorldXInfo world_x_info =
-              ToWorldX(start_us, end_us, inv_time_window, world_start_x, world_width);
-
-          Vec2 pos(world_x_info.world_x_start, world_timer_y);
-          batcher->AddVerticalLine(pos, GetTextBoxHeight(timer_info), z, color,
-                                   std::move(user_data));
-          // For lines, we can ignore the entire pixel into which this event
-          // falls. We align this precisely on the pixel x-coordinate of the
-          // current line being drawn (in ticks). If pixel_delta_in_ticks is
-          // zero, we need to avoid dividing by zero, but we also wouldn't
-          // gain anything here.
-          if (pixel_delta_in_ticks != 0) {
-            min_ignore = min_timegraph_tick +
-                         ((timer_info.start() - min_timegraph_tick) / pixel_delta_in_ticks) *
-                             pixel_delta_in_ticks;
-            max_ignore = min_ignore + pixel_delta_in_ticks;
-          }
-        }
-
-        prev_timer_info = &timer_info;
+        next_text_box = &block[k];
```

Comment:
Could you explain why in the first iteration we are having:
prev = null
current = null
next = block[0]
?
So, why we need to draw then current is the NULLptr?

More precisely, why not work at first with this border case and making the loop starting where current is the first and next is the second (possibly NULL) timer?

Reply:
So this is intentional:
The issue before was, that we tried when looking at the "current" timer, to look forward to the "next" one and messed up the iterator.

The idea is now, to have the current block index pointing to the "next timer" and remember the previous two timers ("previous and current"). So you need no error prone look-ahead logic.

This still means that we will always require n+1 iterations (which is why I extracted the logic into a method).

Also note, that the DrawTimer method will early return if the "current" is nullptr.
So assume we have a just one timer:
Init:
prev = null
current = null
Enter loop:
next = timer
Call Draw:
current == null -> return
Move pointer:
prev = current == nullptr
current = next == timer
Exit the loop:
next = nullptr
Call Draw:
current != null -> Draw the timer

Now assume we have two timers:
Init:
prev = null
current = null
Enter loop iteration 1:
next = timer1
Call Draw:
current == null -> return
Move pointer:
prev = current == nullptr
current = next == timer1
Enter loop iteration 2:
next = timer2
Call Draw:
current == timer1 ->Draw the timer
Move pointer:
prev = current == timer1
current = next == timer2
Exit the loop:
next = nullptr
Call Draw:
current == timer2 -> Draw the timer

Reply:
Added some comments.